File: src/app/services/room.service.ts
import { Injectable } from '@angular/core';
import {Observable} from 'rxjs';
import {Round} from '../model/round';
import {HttpClient} from '@angular/common/http';
import {Player} from '../model/player';
import {Auth} from '../model/auth';
import {UrlService} from './url.service';
import {Room} from '../model/room';
import {PrivateRoom} from '../model/privateRoom';

@Injectable({
  providedIn: 'root'
})
export class RoomService {
  private readonly url: string;

  constructor(private http: HttpClient, private urlService: UrlService) {
    this.url = urlService.roomUrl;
  }

  joinRoom(roomId: number): Observable<Player> {
    return this.http.get<Player>(this.url + '/' + roomId + '/join');
  }

  leaveRoom(id: number): Observable<Player> {
    return this.http.delete<Player>(this.url + '/' + id + '/leave-room');
  }

  getCurrentRound(roomId: number): Observable<Round> {
    return this.http.get<Round>(this.url + '/' + roomId + '/current-round');
  }

  getRooms(): Observable<Room[]> {
    return this.http.get<Room[]>(this.url);
  }

  getRoom(id: number): Observable<Room> {
    return this.http.get<Room>(this.url + '/' + id);
  }

  getPrivateRooms(): Observable<PrivateRoom[]> {
    return this.http.get<PrivateRoom[]>(this.url + '/private');
  }

  getPrivateRoomsFromOwner(): Observable<PrivateRoom[]> {
    return this.http.get<PrivateRoom[]>(this.url + '/private/owner');
  }

  addPrivateRoom(room: PrivateRoom): Observable<PrivateRoom> {
    return this.http.post<PrivateRoom>(this.url + '/private', room);
  }

  addToWhiteList(roomId: number, userId: string): Observable<PrivateRoom> {
    return this.http.patch<PrivateRoom>(this.url + '/private/' + roomId + '/add-user/' + userId, '');
  }

  deleteFromWhiteList(roomId: number, userId: string): Observable<PrivateRoom> {
    return this.http.patch<PrivateRoom>(this.url + '/private/' + roomId + '/remove-user/' + userId, '');
  }

  changeRoom(room: Room): Observable<Room> {
    return this.http.put<Room>(this.url + '/' + room.id, room);
  }

  addRoom(room: Room): Observable<Room> {
    return this.http.post<Room>(this.url, room);
  }

  deleteRoom(room: Room): Observable<Room> {
    return this.http.delete<Room>(this.url + '/' + room.id);
  }

  getPlayer(): Observable<Player> {
    return this.http.get<Player>(this.url + '/players');
  }

  changePlayer(player: Player): Observable<Player> {
    return this.http.put<Player>(this.url + '/players', player);
  }
}

File: app/src/main/java/be/kdg/mobile_client/room/RoomService.java
package be.kdg.mobile_client.room;

import java.util.List;

import be.kdg.mobile_client.room.model.Player;
import io.reactivex.Observable;
import retrofit2.Response;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Path;

/**
 * This service will be used to send API requests to https://poker-game-service.herokuapp.com
 * game micro service back end.
 */
public interface RoomService {
    @GET("/api/rooms")
    Observable<List<Room>> getRooms();

    @GET("/api/rooms/private")
    Observable<List<Room>> getPrivateRooms();

    @GET("/api/rooms/private/owner")
    Observable<List<Room>> getPrivateRoomsOwner();

    @GET("/api/rooms/{id}")
    Observable<Room> getRoom(@Path("id") int id);

    @GET("/api/rooms/{id}/join")
    Observable<Player> joinRoom(@Path("id") int id);

    @DELETE("/api/rooms/{id}/leave-room")
    Observable<Response<Void>> leaveRoom(@Path("id") int id);

    @GET("/api/rooms/{id}/current-round")
    Observable<Response<Void>> getCurrentRound(@Path("id") int id);
}
